### ISSUE TITLE
One context manager with all of the flags?

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#18483]
With the recent announcement on RNGs, I just found out about `jax.threefry_partitionable`.  There are now quite a few flags, each with its own context manager.  I love that they're set with context managers, but they're hard to discover unless you pay really close attention to all of the announcements.  Have you considered adding one context manager for all of the flags?  Something like:
```
def jax_context_manager(enable_x64: bool,
                        enable_custom_prng: bool,
                        threefry_partitionable: bool,
                        cache_dir: str  # currently doesn't use a context manager for some reason
                        ) -> ContextManager[None, None, None]:
    ...
```
This would make it easy to discover flags and encourage users to set them all at the same time.

**Comment by superbobry** [comment]
I agree the current configuration story is not ideal. If the main problem is discoverability, do you think improving the [docs](https://jax.readthedocs.io/en/latest/jax.html#configuration) would help?

On a related note, I started an internal refactoring a few weeks ago, which should eventually translate to meaningful improvements in the public API. The current thinking is

* JAX has two distinct kinds of configuration options: _flags_, set at most once and never changed, and _state objects_, which can be modified any number of times both globally (via `config.update()`) and thread-locally via a context manager.
* We want to clearly separate these on the type level and make the corresponding lookup and update APIs statically typed as well. That should help with catching typos (e.g. `jax.config.enable_custom_rng`) statically and also improve discoverability in editors/IDEs.

**Comment by superbobry** [comment]
I'm not opposed to the idea of having a `config.update()`-like API which updates state objects jointly. I would probably prefer to have just one way of doing thread-local updates, though. So, we would potentially need to deprecate the current context managers in favor of the new API.

I also like that we can get static type errors whenever flags are used for thread-local updates (assuming the changes from my previous message materialize). E..g

```python
# !!! 
with jax.config.thread_local_update(
    jax.config.platforms=["foo", "bar"],
)  
```

**Comment by NeilGirdhar** [comment]
> do you think improving the [docs](https://jax.readthedocs.io/en/latest/jax.html#configuration) would help?

My mistake, I didn't see that they were all collected into section.  That said, what's going on [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.config.html#jax.config)?  Is that `config.update`?

> I'm not opposed to the idea of having a `config.update()`-like API which updates state objects jointly. 

I'm going to use the context manager interface since I do use threads in some invocations, so this wouldn't help me.  I also find context managers more elegant.

> I would probably prefer to have just one way of doing thread-local updates, though. So, we would potentially need to deprecate the current context managers in favor of the new API.

Fair enough.  I personally like smaller interfaces, but I understand that there might be advantages of the current API?
I guess the only other advantage of having one context manager 

> I also like that we can get static type errors

Yeah, I agree!

**Comment by NeilGirdhar** [comment]
Also, `enable_x64` seems to be missing from the configuration section?

**Comment by NeilGirdhar** [comment]
See https://github.com/ray-project/ray/issues/44087#issuecomment-2010753107 for why it would be nice to postpone Jax initialization until a context manager is used.  Right now, you have to very carefully initialize Ray before Jax is imported, which is hard to do in a large project.
